Fallthrough warnings in G++ 7.0.0 experimental

```
[ 85%] Building CXX object example/CMakeFiles/schemavalidator.dir/schemavalidator/schemavalidator.cpp.o
In file included from /home/uruwi/rapidjson/include/rapidjson/schema.h:35:0,
                 from /home/uruwi/rapidjson/example/schemavalidator/schemavalidator.cpp:7:
/home/uruwi/rapidjson/include/rapidjson/internal/regex.h: In member function ‘void rapidjson::internal::GenericRegex<Encoding, Allocator>::Parse(rapidjson::internal::DecodedStream<InputStream, Encoding>&) [with InputStream = rapidjson::GenericStringStream<rapidjson::UTF8<> >; Encoding = rapidjson::UTF8<>; Allocator = rapidjson::CrtAllocator]’:
/home/uruwi/rapidjson/include/rapidjson/internal/regex.h:287:21: error: this statement may fall through [-Werror=implicit-fallthrough=]
                     if (!CharacterEscape(ds, &codepoint))
                     ^~
/home/uruwi/rapidjson/include/rapidjson/internal/regex.h:291:17: note: here
                 default: // Pattern character
                 ^~~~~~~
/home/uruwi/rapidjson/include/rapidjson/internal/regex.h: In member function ‘bool rapidjson::internal::GenericRegex<Encoding, Allocator>::ParseRange(rapidjson::internal::DecodedStream<InputStream, Encoding>&, rapidjson::SizeType*) [with InputStream = rapidjson::GenericStringStream<rapidjson::UTF8<> >; Encoding = rapidjson::UTF8<>; Allocator = rapidjson::CrtAllocator]’:
/home/uruwi/rapidjson/include/rapidjson/internal/regex.h:520:21: error: this statement may fall through [-Werror=implicit-fallthrough=]
                     if (codepoint == '-') {
                     ^~
/home/uruwi/rapidjson/include/rapidjson/internal/regex.h:526:17: note: here
                 case 0:
                 ^~~~
/home/uruwi/rapidjson/include/rapidjson/internal/regex.h:509:17: error: this statement may fall through [-Werror=implicit-fallthrough=]
                 if (ds.Peek() == 'b') {
                 ^~
/home/uruwi/rapidjson/include/rapidjson/internal/regex.h:517:13: note: here
             default:
             ^~~~~~~
cc1plus: all warnings being treated as errors
```

The lines where the warnings direct us already have fallthrough comments, but g++ doesn't seem to recognise them at `-Wimplicit-fallthrough=3`.

--- a/include/rapidjson/internal/regex.h
+++ b/include/rapidjson/internal/regex.h
@@ -29,6 +29,7 @@ RAPIDJSON_DIAG_OFF(implicit-fallthrough)
 #ifdef __GNUC__
 RAPIDJSON_DIAG_PUSH
 RAPIDJSON_DIAG_OFF(effc++)
+RAPIDJSON_DIAG_OFF(implicit-fallthrough)
 #endif
 
 #ifdef _MSC_VER

--- a/include/rapidjson/internal/regex.h
+++ b/include/rapidjson/internal/regex.h
@@ -29,7 +29,9 @@ RAPIDJSON_DIAG_OFF(implicit-fallthrough)
 #ifdef __GNUC__
 RAPIDJSON_DIAG_PUSH
 RAPIDJSON_DIAG_OFF(effc++)
+#if __GNUC__ >= 7
 RAPIDJSON_DIAG_OFF(implicit-fallthrough)
+#endif
 #endif
 
 #ifdef _MSC_VER

